docs/C_API_Guidelines.md

```diff
@@ -0,0 +1,83 @@
+
+Cleanup API that destroys objects or simply deallocates memory must return void. Most of the time such API can never error out. Adding return status creates more uncertainty for the client and does not help in exception scenarios such as try/finally in C#. Returning void helps clients to write cleaner code and preserve original exception if any with its meaningful error message rather than memory deallocation failure.
+
+This requirement will also help us to create C++ API wrappers that are exception safe.
+
+Consider logging errors if you must rather than return them to the client.  
+
+Example: on Windows delete operator is implemented on top of HeapFree() which may return an error. However, delete never returns anything and can be relied upon as a no throw primitive for cleanup purposes.
+
+### 5. Public API must not require calling code to cleanup anything when it errors out
+
+When API errors out it must leave all its out parameters and buffers untouched, in its original condition. All memory allocations must be cleaned up and no memory leaks result.
+
+The obvious exception in this rule is the actual OrtStatus that is dynamically allocated and must be released by the client using the corresponding API.
+
+Some of the client code, notably in C#, attempts to detect which out arguments need a cleanup when an API errors out. The way it is done, out arguments are pre-set to a specific value, such as zero. If the API errors out, the client code attempts to cleanup if the out argument has changed.  
+
+Such a technique is error prone and dangerous, as the client has no way of finding out if the out argument has already been cleaned up by the API as should be the case. It may result in double free. One reason for this is our insufficient documentation. This also results in a convoluted hard to read code with nested try/finally/catch clauses.
+
+It seems that most of our API is compliant with it. Some API zero out the out arguments right away. It is fine.
+
+Examples of an API that are compliant with this requirement are: GetBoundOutputNames and GetBoundOutputValues.
+
+### 6. Public API that allocates memory must take an allocator parameter to use during the allocation
+
+APIs that require memory allocation to return results, must take the instance of an OrtAllocator to use for such allocations. This serves two purposes:
+
+* The user may want to supply their own allocator to use. Many of our APIs do that.
+
+* The API does not have to declare a separate entry point for deallocating memory specifically for its type of allocation as we already have such entry points. OrtAllocatorAlloc/OrtAllocatorFree.
+
+Consider established patterns when APIs return multiple allocations.
+
+### 7. Public API must document that all strings they accept, and return are UTF-8 encoded
+
+All API return and accept strings in UTF-8 encodings. We must be mindful of that when maintaining language bindings.
```

Comment:
```suggestion
All APIs must return and accept strings in UTF-8 encodings. We must be mindful of that when maintaining language bindings.
```

Reply:
The CreateSession API still uses wchar on Windows for the path right? It's UTF-8 on macOS and Linux.